feat(fuse): Expose MFS as FUSE mount point

gsergey418 opened the pull request:
- Fixes #10710
- Fixes #7738
- Fixes #5111
- Partially addresses need from https://github.com/ipfs/kubo/issues/5504 (for MFS, not IPNS)

Expose the MFS root as a FUSE filesystem mounted at /mfs. 
<!--
Please update docs/changelogs/ if you're modifying Go files. If your change does not require a changelog entry, please do one of the following:
- add `[skip changelog]` to the PR title
- label the PR with `skip/changelog`
-->

gsergey418 commented:
Core functionality should be ready, need to write some tests and change up the documentation.

gsergey418 commented:
Tests ready now, PR should be good for review.

guillaumemichel commented on docs/changelogs/v0.35.md:
```diff
@@ -42,6 +43,10 @@ The WebUI, accessible at http://127.0.0.1:5001/webui/, now includes support for
 
+#### Expose MFS as a FUSE mount point
+
+The MFS root is now available as a read/write FUSE mount point at `/mfs`. This filesystem is mounted in the same way as `/ipfs` and `/ipns` when running `ipfs mount` or `ipfs daemon --mount`.
```
Do `writes` to MFS actually work? I wasn't able to test.

gsergey418 replied:
Yeah, cp, touch, mkdir commands work as expected. (on my machine) I don't think you should take my word for it tho.

guillaumemichel commented on fuse/node/mount_unix.go:
```diff
@@ -43,15 +44,19 @@ func Mount(node *core.IpfsNode, fsdir, nsdir string) error {
 }
 
-func doMount(node *core.IpfsNode, fsdir, nsdir string) error {
+func doMount(node *core.IpfsNode, fsdir, nsdir, mfdir string) error {
```
```suggestion
func doMount(node *core.IpfsNode, fsdir, nsdir, mfsdir string) error {
```

guillaumemichel commented on fuse/node/mount_unix.go:
```diff
@@ -67,8 +72,8 @@ func doMount(node *core.IpfsNode, fsdir, nsdir string) error {
 	// this sync stuff is so that both can be mounted simultaneously.
-	var fsmount, nsmount mount.Mount
-	var err1, err2 error
+	var fsmount, nsmount, mfmount mount.Mount
```
```suggestion
	var fsmount, nsmount, mfsmount mount.Mount
```

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,349 @@
+}
+
+// Check that our structs implement all the interfaces we want.
+type mfDir interface {
```
```suggestion
type mfsDir interface {
```

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,349 @@
+
+var _ mfDir = (*Dir)(nil)
+
+type mfFile interface {
```
```suggestion
type mfsFile interface {
```

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,349 @@
+
+var _ mfFile = (*File)(nil)
+
+type mfHandler interface {
```
```suggestion
type mfsHandler interface {
```

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,349 @@
+}
+
+// Sync the file's contents to MFS.
+func (file *File) Fsync(ctx context.Context, req *fuse.FsyncRequest) error {
```
is it used anywhere?

gsergey418 replied:
Pretty sure it is used when writing a file to the disk

guillaumemichel reviewed (commented):
I didn't review everything at the moment.

@gsergey418 were you able to test mounting `/mfs` and interact with that directory (read/write)? 

I didn't manage to test myself, simply mounting `/ipfs` and `/ipns` (using master) broke fuse in my debian VM test setup.

Would you mind updating [fuse.md](https://github.com/ipfs/kubo/blob/master/docs/fuse.md) to reflect these changes?

gsergey418 commented:
> I didn't review everything at the moment.
> 
> @gsergey418 were you able to test mounting `/mfs` and interact with that directory (read/write)?
> 
> I didn't manage to test myself, simply mounting `/ipfs` and `/ipns` (using master) broke fuse in my debian VM test setup.
> 
> Would you mind updating [fuse.md](https://github.com/ipfs/kubo/blob/master/docs/fuse.md) to reflect these changes?

@guillaumemichel Yeah, it mounts and umounts successfully on my machine and I can interact with it like a normal filesystem (ls, cat, touch, mkdir, stat, etc...). Reading and writing seems to work correctly and keeps the files intact. I've changed the variable names like you requested and added documentation.

lidel commented:
Triage note:  thank you, due to holiday season, we may be slower to respond, will look into this after tackling priority work for 0.35.0-rc1

guillaumemichel commented:
Thanks @gsergey418 for the PR.

I have tested this branch on a fresh Debian Bookworm install, and it seems that most operations work great in fuse. E.g files added with `ipfs add --to-files=<filename>` can be listed from fuse, and files added from fuse can be listed with `ipfs files`.

> I think it would be good to add some way of displaying the file's CID, perhaps through xattrs, if you want full MFS functionality.

I didn't manage to display a file's CID using xattrs
```sh
$ xattr /mfs/myDir/hello.txt
[Errno 95] Operation not supported: b'/mfs/myDir/hello.txt'
$ xattr /home/guissou/.ipfs/config
$ ...
```
The latest command works on a file outside of MFS (but I guess the file doesn't have any xattr attribute).

Is this feature working yet?

---

It seems that in the mounted MFS (fuse), files are readonly. It is possible to override a file, but not to edit it. It that expected? What would it take for files to be mutable?

gsergey418 commented:
> Thanks @gsergey418 for the PR.
> 
> I have tested this branch on a fresh Debian Bookworm install, and it seems that most operations work great in fuse. E.g files added with `ipfs add --to-files=<filename>` can be listed from fuse, and files added from fuse can be listed with `ipfs files`.
> 
> > I think it would be good to add some way of displaying the file's CID, perhaps through xattrs, if you want full MFS functionality.
> 
> I didn't manage to display a file's CID using xattrs
> 
> ```shell
> $ xattr /mfs/myDir/hello.txt
> [Errno 95] Operation not supported: b'/mfs/myDir/hello.txt'
> $ xattr /home/guissou/.ipfs/config
> $ ...
> ```
> 
> The latest command works on a file outside of MFS (but I guess the file doesn't have any xattr attribute).
> 
> Is this feature working yet?
> 
> It seems that in the mounted MFS (fuse), files are readonly. It is possible to override a file, but not to edit it. It that expected? What would it take for files to be mutable?

@guillaumemichel Thank you for the review. I will look into it.

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,395 @@
+
+// Directory attributes (stat).
+func (dir *Dir) Attr(ctx context.Context, attr *fuse.Attr) error {
+	attr.Mode = os.FileMode(os.ModeDir | 0755)
+	attr.Size = 4096
+	attr.Blocks = 8
```
What is the reasoning for using these magic numbers? Could we link to already existing consts instead?

guillaumemichel commented on fuse/mfs/mfs_unix.go:
```diff
@@ -0,0 +1,395 @@
+		return err
+	}
+	attr.Size = uint64(size)
+	if size%512 == 0 {
+		attr.Blocks = uint64(size / 512)
+	} else {
+		attr.Blocks = uint64(size/512 + 1)
```
what does `512` correspond to?

gsergey418 replied:
Standard block size on the filesystem

guillaumemichel reviewed (commented):
It works great!

@gsergey418 could you also add some [sharness](https://github.com/ipfs/kubo/tree/master/test/sharness) tests to make sure `/mfs` works as expected (e.g by adding a file using `ipfs add --to-files` and verify that it appears in `/mfs`, copying a file using `cp myFile /mfs/myFile` and verifying with `ipfs files ls` that the file exists). It would be great to test all commands (`touch`, `ls`, `mkdir`, `cp`, `mv`, `rm`), and ideally also the xattrs.

Let us know if you have issues with dependencies in the tests setup (`fuse`, `xattrs`)

gsergey418 commented:
> It seems that in the mounted MFS (fuse), files are readonly. It is possible to override a file, but not to edit it. It that expected? What would it take for files to be mutable?

I've noticed some problems with setting file permission which is done by some programs like sed and vim, which was most likely causing the immutability problem. Another problem with the file not being overwritten on mv. I'll fix those and let you know.

guillaumemichel commented:
@gsergey418 thanks for the investigation.

It makes sense, we can leave it as is and inform users in the appropriate docs about what isn't supported yet.

It seems like the long term solution would be to edit MFS directly to support file metadata modifications, but we can tackle it later.

> Anyways, have a good weekend

You too :smile:

guillaumemichel commented:
Triage notes:
* We plan to include this PR in v0.36
* @gsergey418 could you add some sharness tests to test basic functionality. Some tests could be
  1. `touch`, `echo >file.txt`, or `cp` content in the fuse mountpoint, and verify that the content reads as expected with `ipfs files`
  2. add content to MFS using `ipfs add --to-files` and verify that the content appears in the fuse mountpoint

gsergey418 commented:
@guillaumemichel I've added some tests, but haven't run them yet, can't get it to work on my machine. I'll check it out in CI.

guillaumemichel commented:
It seems that the FUSE sharness tests haven't been run by the CI either.

gsergey418 commented:
I think you need to set the `TEST_FUSE` env var to `1` in your CI environment.

guillaumemichel commented:
@galargh do you know how to set `TEST_FUSE=1` to enable FUSE sharness tests?

gsergey418 commented:
Let's see if this does anything. t0300-mount.sh AFAIK is the only FUSE sharness test, I don't think it's worth inventing something new just to test this.

gsergey418 commented:
I think that broke some other tests in different scripts, so maybe they would have to be fixed or removed. I'm going on holiday rn, so I've won't be working on this for the next few days

gsergey418 commented:
I couldn't get the FUSE tests in CI working, so probably they were disabled for a reason. I'm still going to keep those sharness tests and lib changes just in case though.

gsergey418 commented:
@guillaumemichel  No, they don't work either way

guillaumemichel reviewed (approved):
OK! Thanks a lot for your contribution @gsergey418!

I am good with merging the PR. FUSE is currently an experimental feature, and we can iterate from this initial state if we want to improve MFS later on to support attribute storage for example.

@lidel do we want this PR in v0.35.0? If not we can wait before merging

gsergey418 commented:
Awesome! :D

lidel commented:
## my todo

- [x] resolve merge conflicts
- [x] add warnings about mutable endpoint risks (undesired preload of lazy-loaded DAGs, flaky writes not ready for production)
- [ ] see if we can easily fix sharness to pass run with `TEST_FUSE=1`
- [ ] merge

lidel reviewed (approved):
Thank you. Did some testing and while writes are flakey, read-only works good enough for 0.35.0-rc1.

To set user expectations, added some caveats and warnings about experimental nature of this feature and fixed small bugs that broke `TEST_FUSE=1 ./t0030-mount.sh -v` sharness tests. They pass now.

Merging :+1:
